src/backend/distributed/planner/multi_join_order.c

```diff
@@ -781,17 +797,38 @@ static JoinOrderNode *
 ReferenceJoin(JoinOrderNode *currentJoinNode, TableEntry *candidateTable,
 			  List *applicableJoinClauses, JoinType joinType)
 {
-	JoinOrderNode *nextJoinNode = NULL;
 	int applicableJoinCount = list_length(applicableJoinClauses);
-	char candidatePartitionMethod = PartitionMethod(candidateTable->relationId);
-	char leftPartitionMethod = PartitionMethod(currentJoinNode->tableEntry->relationId);
-	bool performReferenceJoin = false;
-
 	if (applicableJoinCount <= 0)
 	{
 		return NULL;
 	}
 
+	JoinOrderNode *nextJoinNode =
+		CartesianProductReferenceJoin(currentJoinNode, candidateTable,
```

Comment:
Can we refactor this so that `ReferenceJoin` is not calling to `CartesianProductReferenceJoin`? It is a bit strange to read and understand the logic here.

Instead I would reuse `IsReferenceJoin` to derive `performReferenceJoin` and call the correct `MakeJoinOrderNode` call.